ax25tools/man/axparms.8
.TH AXPARMS 8 "19 January 2017" Linux "Linux System Managers Manual"
.SH NAME
axparms \- Configure AX.25 interfaces.
.SH SYNOPSIS
.B axparms --assoc|--forward|--route|--setcall|--version ...
.SH DESCRIPTION
.LP
The
.B axparms
command is designed to be a multi-function command that allows miscellaneous
commands to be issued to the Linux AX.25 protocol layer.  The different modes
of the command are chosen by the first argument. Subsequent arguments depend
upon this argument and so no generalised command format can be given.
.SH "--assoc Argument"
.LP
The format of this option is:
.LP
.nf
.B axparms --assoc <callsign> <username|uid>
.br
.B axparms --assoc <callsign> delete
.br
.B axparms --assoc policy [default|deny]
.br
.B axparms --assoc show
.fi
.LP
This option manipulates the kernel uid/callsign mapping table, allowing
callsigns to be associated and dis-associated with a user. The
.B policy
option permits the superuser to have all other uid's either default to the
actual port name, or to block traffic.
.LP
At power up the table is blank and the policy is 'default', which is thus
backward compatible.
.SH "--forward Argument"
.LP
Allows the use of many receivers with one transmitter, known as packet
forwarding in many systems. The format of this command is:
.LP
.nf
.B axparms --forward <portfrom> <portto>
.br
.B axparms --forward <portfrom> delete
.fi
.LP
Any packets to be transmitted on port portfrom will be transmitted on port
portto. This will stay in force until the second form of the command is
issued which will remove the association.
.SH "--route Argument"
.LP
This option allows the internal AX.25 routing table to be manipulated. This
table is available for reading in /proc/net/ax25_route, and will be built up
dynamically by stations heard. However it is possible to add, delete and list
entries via this option.
.LP
The formats of this option are:
.LP
.nf
.B axparms --route add <port> <callsign> [<digis>] [--ipmode V|D]
.br
.B axparms --route del <port> <callsign>
.fi
.B axparms --route list
.fi
.LP
Routes added via this command will not be removed from the internal routing
table when they are \(lqold\(rq as normal entries are. The
.B --ipmode
option sets mode vc or mode datagram for this destination.
.LP
If the <callsign> argument is set to \(lqdefault\(rq then this will set the
default route for all outgoing AX.25 connections which will be used when there
is no specific route to the required destination.
.SH "--setcall Argument"
.LP
The format of this option is:
.LP
.B axparms --setcall <interface> <callsign>
.LP
This changes the callsign associated at the given physical ax25 interface.
.LP
Cave: The interface name is not the symbolic port name from axports, but
the real interface name (from ifconfig(8)):
ax0, ax1, .., sp0, .., bpq0, ... etc..
.LP
The change is permanent as long as the interface exists, or another \(lqaxparms
--setcall\(rq is issued.
.LP
The interface has to exist already in order to use this option; it
may be in state UP or DOWN.
.SH "--version Argument"
.LP
This option displays the version of the AX.25 utilities that
.B axparms
belongs to.
.SH FILES
.LP
/proc/net/bpqether
.br
/proc/net/ax25_calls
.br
/etc/ax25/axports
.SH "SEE ALSO"
.BR call (1),
.BR getsockopt (2),
.BR setsockopt (2),
.BR ax25 (4),
.BR axctl (8),
.BR axports (5).
.SH AUTHORS
.nf
Alan Cox GW4PTS <alan@cymru.net>
.br
Jonathan Naylor G4KLX <g4klx@g4klx.demon.co.uk>
.br
Joerg Reuter DL1BKE <jreuter@poboxes.com>
.fi


ax25apps/NEWS
NEWS for ax25-apps
==================

These programs are still being tested, you should expect some quirks
especially with compiling for a little while longer




ax25tools/ax25/axparms.c
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <pwd.h>

#include <config.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>

#include <net/if.h>

#include <netax25/ax25.h>
#include <netrose/rose.h>

#include <netax25/axlib.h>
#include <netax25/axconfig.h>

#include "../pathnames.h"

static void usage(void)
{
	fprintf(stderr, "usage: axparms --assoc|--forward|--route|--setcall|--version ...\n");
}

static void usageassoc(void)
{
	fprintf(stderr, "usage: axparms --assoc show\n");
	fprintf(stderr, "usage: axparms --assoc policy default|deny\n");
	fprintf(stderr, "usage: axparms --assoc [callsign] [username]\n");
	fprintf(stderr, "usage: axparms --assoc [callsign] delete\n");
}

static void usageforward(void)
{
	fprintf(stderr, "usage: axparms --forward <portfrom> <portto>\n");
	fprintf(stderr, "usage: axparms --forward <portfrom> delete\n");
}

static void usageroute(void)
{
	fprintf(stderr, "usage: axparms --route add port callsign [digi ...] [--ipmode mode]\n");
	fprintf(stderr, "usage: axparms --route del port callsign\n");
	fprintf(stderr, "usage: axparms --route list\n");
}

static void usagesetcall(void)
{
	fprintf(stderr, "usage: axparms --setcall interface callsign\n");
}

static int routes(int s, int argc, char *argv[], ax25_address *callsign)
{
	struct ax25_routes_struct ax25_route;
	struct ax25_route_opt_struct ax25_opt;
	int i, j;
	int ip_mode = ' ';
	FILE* fp;
	char routebuf[80];

	if (strcmp(argv[2], "add") == 0) {
		ax25_route.port_addr  = *callsign;
		ax25_route.digi_count = 0;

		if (strcmp(argv[4], "default") == 0) {
			ax25_route.dest_addr = null_ax25_address;
		} else {
			if (ax25_aton_entry(argv[4], (char *)&ax25_route.dest_addr) == -1)
				return 1;
		}

		for (i = 5, j = 0; i < argc && j < 6; i++) {
			if (strncmp(argv[i], "--i", 3) == 0 || strncmp(argv[i], "-i", 2) == 0) {
				if (++i == argc) {
					fprintf(stderr, "axparms: -i must have a parameter\n");
					return 1;
				}
				switch (*argv[i]) {
				case 'd':
				case 'D':
					ip_mode = 'D';
					break;
				case 'v':
				case 'V':
					ip_mode = 'V';
					break;
				default:
					ip_mode = ' ';
					break;
				}
			} else {
				if (ax25_aton_entry(argv[i], (char *)&ax25_route.digi_addr[j]) == -1)
					return 1;
				ax25_route.digi_count++;
				j++;
			}
		}

		if (ioctl(s, SIOCADDRT, &ax25_route) != 0) {
			perror("axparms: SIOCADDRT");
			return 1;
		}

		ax25_opt.port_addr = *callsign;
		ax25_opt.dest_addr = ax25_route.dest_addr;
		ax25_opt.cmd = AX25_SET_RT_IPMODE;
		ax25_opt.arg = ip_mode;

		if (ioctl(s, SIOCAX25OPTRT, &ax25_opt) != 0) {
			perror("axparms: SIOCAX25OPTRT");
			return 1;
		}
	}

	if (strcmp(argv[2], "del") == 0) {
		ax25_route.port_addr  = *callsign;
		ax25_route.digi_count = 0;

		if (strcmp(argv[4], "default") == 0) {
			ax25_route.dest_addr = null_ax25_address;
		} else {
			if (ax25_aton_entry(argv[4], (char *)&ax25_route.dest_addr) == -1)
				return 1;
		}

		if (ioctl(s, SIOCDELRT, &ax25_route) != 0) {
			perror("axparms: SIOCDELRT");
			return 1;
		}
	}

	if (strcmp(argv[2], "list") == 0) {
		fp = fopen(PROC_AX25_ROUTE_FILE, "r");
		if (fp == NULL) {
			fprintf(stderr, "axparms: route: cannot open %s\n",
PROC_AX25_ROUTE_FILE);
			return 1;
		}
		while (fgets(routebuf,80,fp))
			printf("%s", routebuf);
		puts("");
	}

	return 0;
}

static int setifcall(int s, char *ifn, char *name)
{
	char call[7];
	struct ifreq ifr;

	if (ax25_aton_entry(name, call) == -1)
		return 1;

	strcpy(ifr.ifr_name, ifn);
	memcpy(ifr.ifr_hwaddr.sa_data, call, 7);
	ifr.ifr_hwaddr.sa_family = AF_AX25;

	if (ioctl(s, SIOCSIFHWADDR, &ifr) != 0) {
		perror("axparms: SIOCSIFHWADDR");
		return 1;
	}

	return 0;
}

static int associate(int s, int argc, char *argv[])
{
	char buffer[80], *u, *c, *endp;
	struct sockaddr_ax25 sax25;
	struct passwd *pw;
	uid_t uid;
	int opt;
	FILE *fp;


	if (strcmp(argv[2], "show") == 0) {
		if (argc < 3) {
			usageassoc();
			exit(1);
		}

		fp = fopen(PROC_AX25_CALLS_FILE, "r");
		if (fp == NULL) {
			fprintf(stderr, "axparms: associate: cannot open %s\n", PROC_AX25_CALLS_FILE);
			return 1;
		}

		fgets(buffer, 80, fp);

		printf("Userid     Callsign\n");

		while (fgets(buffer, 80, fp) != NULL) {
			u = strtok(buffer, " \t\n");
			c = strtok(NULL, " \t\n");
			pw = getpwuid(atoi(u));
			if (pw != NULL)
				printf("%-10s %s\n", pw->pw_name, c);
		}

		fclose(fp);

		return 0;
	}

	if (strcmp(argv[2], "policy") == 0) {
		if (argc < 4) {
			usageassoc();
			exit(1);
		}

		if (strcmp(argv[3], "default") == 0) {
			opt = AX25_NOUID_DEFAULT;

			if (ioctl(s, SIOCAX25NOUID, &opt) == -1) {
				perror("axparms: SIOCAX25NOUID");
				return 1;
			}

			return 0;
		}

		if (strcmp(argv[3], "deny") == 0) {
			opt = AX25_NOUID_BLOCK;

			if (ioctl(s, SIOCAX25NOUID, &opt) == -1) {
				perror("axparms: SIOCAX25NOUID");
				return 1;
			}

			return 0;
		}

		fprintf(stderr, "axparms: associate: 'default' or 'deny' required\n");

		return 1;
	}

	if (argc < 4) {
		usageassoc();
		exit(1);
	}

	if (ax25_aton_entry(argv[2], (char *)&sax25.sax25_call) == -1) {
		fprintf(stderr, "axparms: associate: invalid callsign %s\n", argv[2]);
		return 1;
	}

	if (strcmp(argv[3], "delete") == 0) {
		if (ioctl(s, SIOCAX25DELUID, &sax25) == -1) {
			perror("axparms: SIOCAX25DELUID");
			return 1;
		}

		return 0;
	}

	/*
	 * Tolerate spaces following a UID, it may happen in scripts
	 */
	errno = 0;
	uid = strtol(argv[3], &endp, 0);
	if (!errno && (*endp == '\0' || isspace(*endp))) {
		sax25.sax25_uid = uid;
	} else {
		pw = getpwnam(argv[3]);

		if (pw == NULL) {
			fprintf(stderr,
				"axparms: associate: unknown username %s\n",
				argv[3]);
			return 1;
		}

		sax25.sax25_uid = pw->pw_uid;
	}

	if (ioctl(s, SIOCAX25ADDUID, &sax25) == -1) {
		perror("axparms: SIOCAX25ADDUID");
		return 1;
	}

	return 0;
}

static int forward(int s, int argc, char *argv[])
{
#ifdef HAVE_AX25_FWD_STRUCT
	struct ax25_fwd_struct ax25_fwd;
	char *addr;

	if (argc < 4) {
		usageforward();
		exit(1);
	}

	if (ax25_config_load_ports() == 0) {
		fprintf(stderr, "axparms: no AX.25 port data configured\n");
		return 1;
	}

	addr = ax25_config_get_addr(argv[2]);
	if (addr == NULL) {
		fprintf(stderr, "axparms: invalid port name - %s\n", argv[2]);
		return 1;
	}

	if (ax25_aton_entry(addr, (char *)&ax25_fwd.port_from) == -1) {
		fprintf(stderr, "axparms: invalid port name - %s\n", argv[2]);
		return 1;
	}

	if (strcmp(argv[3], "delete") == 0) {
		if (ioctl(s, SIOCAX25DELFWD, &ax25_fwd) == -1) {
			perror("axparms: SIOCAX25DELFWD");
			return 1;
		}

		return 0;
	}

	addr = ax25_config_get_addr(argv[3]);
	if (addr == NULL) {
		fprintf(stderr, "axparms: invalid port name - %s\n", argv[3]);
		return 1;
	}

	if (ax25_aton_entry(addr, (char *)&ax25_fwd.port_to) == -1) {
		fprintf(stderr, "axparms: invalid port name - %s\n", argv[3]);
		return 1;
	}

	if (ioctl(s, SIOCAX25ADDFWD, &ax25_fwd) == -1) {
		perror("axparms: SIOCAX25ADDFWD");
		return 1;
	}
#else
	fprintf(stderr, "axparms: Not compiled in with forwarding option.\n");
#endif /* HAVE_AX25_FWD_STRUCT */

	return 0;
}

int main(int argc, char **argv)
{
	ax25_address callsign;
	int s, n;
	char *addr;

	if (argc == 1) {
		usage();
		return 1;
	}

	if (strncmp(argv[1], "--v", 3) == 0 || strncmp(argv[1], "-v", 2) == 0) {
		printf("axparms: %s\n", VERSION);
		return 0;
	}

	if (strncmp(argv[1], "--a", 3) == 0 || strncmp(argv[1], "-a", 2) == 0) {

		if (argc < 3) {
			usageassoc();
			return 1;
		}

		s = socket(AF_AX25, SOCK_SEQPACKET, 0);
		if (s < 0) {
			perror("axparms: socket");
			return 1;
		}

		n = associate(s, argc, argv);

		close(s);

		return n;
	}

	if (strncmp(argv[1], "--f", 3) == 0 || strncmp(argv[1], "-f", 2) == 0) {
		if (argc == 2) {
			usageforward();
			return 1;
		}

		s = socket(AF_AX25, SOCK_SEQPACKET, 0);
		if (s < 0) {
			perror("axparms: socket");
			return 1;
		}

		n = forward(s, argc, argv);

		close(s);

		return n;
	}

	if (strncmp(argv[1], "--r", 3) == 0 || strncmp(argv[1], "-r", 2) == 0) {
		if (argc < 3 ) {
			usageroute();
			return 1;
		}

		if (strcmp(argv[2], "add") != 0 && strcmp(argv[2], "del") != 0 && strcmp(argv[2], "list") != 0) {
			usageroute();
			return 1;
		}

		if (argc < 5 && strcmp(argv[2], "list") != 0) {
			usageroute();
			return 1;
		}

		if (ax25_config_load_ports() == 0) {
			fprintf(stderr, "axparms: no AX.25 port data configured\n");
			return 1;
		}

		addr = ax25_config_get_addr(argv[3]);
		if (addr == NULL) {
			fprintf(stderr, "axparms: invalid port name - %s\n", argv[3]);
			return 1;
		}

		if (ax25_aton_entry(addr, callsign.ax25_call) == -1)
			return 1;

		s = socket(AF_AX25, SOCK_SEQPACKET, 0);
		if (s < 0) {
			perror("axparms: socket");
			return 1;
		}

		n = routes(s, argc, argv, &callsign);

		close(s);

		return n;
	}

	if (strncmp(argv[1], "--s", 3) == 0 || strncmp(argv[1], "-s", 2) == 0) {
		if (argc != 4) {
			usagesetcall();
			return 1;
		}

		s = socket(AF_INET, SOCK_DGRAM, 0);
		if (s < 0) {
			perror("axparms: socket");
			return 1;
		}

		n = setifcall(s, argv[2], argv[3]);

		close(s);

		return n;
	}

	usage();

	return 1;
}
